[pytorch]Migrate _th_ger to Aten and kill resize_scalar in codegen

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33792 [pytorch]Migrate _th_ger to Aten and kill resize_scalar in codegen**

Differential Revision: [D20107158](https://our.internmc.facebook.com/intern/diff/D20107158)

Commit:
[pytorch]Migrate _th_ger to Aten and kill resize_scalar in codegen

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -156,6 +156,21 @@ Tensor& addr_out(Tensor &result, const Tensor& self, const Tensor& vec1, const T
+  if (result.numel() == 0) {
+    result.resize_({ self.size(0), vec2.size(0) });
+  }
+  return at::_addr_out(result, result, self, vec2, Scalar(0), Scalar(1));
```
This doesn't seem right: `at::_addr_out(result, result, self, vec2, ...)` if I understand it correctly does:
result = (result + outer_product(self, vec2)). What we really want here is `result = outer_product(self, vec2)`, right?

zou3519 replied:
Nvm, I see. The beta=0 makes it so that the addition is irrelevant

zou3519 commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -156,6 +156,21 @@ Tensor& addr_out(Tensor &result, const Tensor& self, const Tensor& vec1, const T
+Tensor& ger_out(Tensor &result, const Tensor& self, const Tensor& vec2) {
+  check_1d(self, "self", "ger");
+  check_1d(vec2, "vec2", "ger");
+  if (result.numel() == 0) {
```
This change is BC breaking --
```
x = torch.randn(5)
y = torch.randn(7)
out = torch.randn(2, 3)
torch.ger(x, y, out=out) # used to be OK.
```

If we want to change the semantics we should do a deprecation cycle by allowing the old behavior (resize anything to the correct size) but raise a deprecation warning.

glaringlee replied:
Ah, I see. I will support the old behavior for now but print out a warning letting user switch to this new convention.

glaringlee replied:
back to old behavior as we discussed

zou3519 commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -1386,11 +1386,7 @@ def resize_arg(arg):
         if isinstance(resize, str):
             return "{}.resize_({}.sizes());".format(arg['name'], resize)
         else:
-            resize_scalar = arg.get('resize_scalar', False)
```
The existence of `resize_scalar` makes it seem like ger, at some point in the past, supported accepting a 0D tensor. It doesn't accept supporting 0D tensors on master and no one has complained about it, so this change seems fine to me.

If we want to be really safe we can try to figure out if `torch.ger` accepted a 0D tensor at any time in the past and figure out when `torch.ger` stopped accepting a 0D tensor, if at all.

glaringlee replied:
Checked a little bit. I think people who made the _th_ger change wanted to make this resize safe.
In legacy code, the size check is inside addr function, but this resize happens before calling addr. _th_ger call addr underline, and addr doesn't allow 0D vec. So ger doesn't support 0D vec anyway

zou3519 reviewed (changes requested):
The new resize semantics introduced in this PR are bc breaking; we should either:
1) follow the old behavior
2) follow the old behavior and issue a deprecation warning.

Commit:
Update on "[pytorch]Migrate _th_ger to Aten and kill resize_scalar in codegen"


Differential Revision: [D20107158](https://our.internmc.facebook.com/intern/diff/D20107158)

[ghstack-poisoned]

glaringlee commented:
discussed with @zou3519 , we will keep the old TH resize behavior in this PR. We will open a new PR if we need to deprecated the legacy resize behavior.

gchanan commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -156,6 +156,21 @@ Tensor& addr_out(Tensor &result, const Tensor& self, const Tensor& vec1, const T
+Tensor& ger_out(Tensor &result, const Tensor& self, const Tensor& vec2) {
+  check_1d(self, "self", "ger");
+  check_1d(vec2, "vec2", "ger");
+  if (result.dim() != 2 || result.size(0) != self.size(0) || result.size(1) != vec2.size(0)) {
```
nit: I think it's clearer if you do something like
```
if (result.sizes() != {self.size(0), vec2.size(0)}) {
   result.resize_({...});
}
```

glaringlee replied:
Will do if there are more dimensions to check. For 2 dimensions, I think this is still fine and save a memory allocation, faster.

Commit:
Update on "[pytorch]Migrate _th_ger to Aten and kill resize_scalar in codegen"


Differential Revision: [D20107158](https://our.internmc.facebook.com/intern/diff/D20107158)

[ghstack-poisoned]

glaringlee commented:
This breaks XLA CI test, added Ailing to give update here once XLA side is ready

ailzhang commented:
@pytorchbot rebase this please

ailzhang commented:
ehhh what happened to our pytorchbot? :P 
@glaringlee I think this PR should be fine if you rebase on top of master. Would you mind rebasing and see whether XLA test pass?  
The current failure is caused by my update of our API between PT and XLA.

ailzhang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -156,6 +156,21 @@ Tensor& addr_out(Tensor &result, const Tensor& self, const Tensor& vec1, const T
+
+Tensor ger(const Tensor& self, const Tensor& vec2) {
+  Tensor result = at::empty({0}, self.options());
+  at::native::ger_out(result, self, vec2);
```
In general it's good to call `at::ger_out` instead of `at::native`. It doesn't matter in the current implementation, but it might cause trouble for XLA if we change `ger_out` to be non-compound op in the future.

glaringlee replied:
changed to at::

Commit:
Update on "[pytorch]Migrate _th_ger to Aten and kill resize_scalar in codegen"


Differential Revision: [D20107158](https://our.internmc.facebook.com/intern/diff/D20107158)

[ghstack-poisoned]

Commit:
Update on "[pytorch]Migrate _th_ger to Aten and kill resize_scalar in codegen"


Differential Revision: [D20107158](https://our.internmc.facebook.com/intern/diff/D20107158)

[ghstack-poisoned]

ailzhang reviewed (approved):
Thanks!

Commit:
Update on "[pytorch]Migrate _th_ger to Aten and kill resize_scalar in codegen"


Differential Revision: [D20107158](https://our.internmc.facebook.com/intern/diff/D20107158)

[ghstack-poisoned]

Commit:
Update on "[pytorch]Migrate _th_ger to Aten and kill resize_scalar in codegen"


Differential Revision: [D20107158](https://our.internmc.facebook.com/intern/diff/D20107158)

[ghstack-poisoned]
